lib/faktory_worker/worker.ex

```diff
@@ -69,9 +127,57 @@ defmodule FaktoryWorker.Worker do
 
   defp schedule_beat(%{worker_state: worker_state} = state)
        when worker_state in @valid_beat_states do
-    beat_ref = Process.send_after(state.worker_server, :beat, state.beat_interval)
+    beat_ref = Process.send_after(self(), :beat, state.beat_interval)
     %{state | beat_ref: beat_ref}
   end
 
   defp schedule_beat(state), do: state
+
+  defp handle_fetch_response({{:ok, :no_content}, conn}, state) do
+    %{state | conn: conn}
+  end
+
+  defp handle_fetch_response({{:ok, job}, conn}, state) do
+    job_supervisor =
+      state.worker_config
+      |> faktory_name()
+      |> FaktoryWorker.JobSupervisor.format_supervisor_name()
+
+    job_ref =
+      Task.Supervisor.async_nolink(
+        job_supervisor,
+        state.worker_module,
+        :perform,
+        job["args"],
+        # make this whatever is set for the 'reserve_for' minus 20 seconds
+        shutdown: :infinity
+      )
+
+    %{state | conn: conn, worker_state: :running_job, job_ref: job_ref, job_id: job["jid"]}
+  end
+
+  defp schedule_fetch(%{disable_fetch: true} = state), do: state
+
+  defp schedule_fetch(%{worker_state: worker_state} = state) when worker_state == :ok do
+    :ok = Process.send(self(), :fetch, [])
+    state
+  end
+
+  defp schedule_fetch(state), do: state
+
+  defp handle_ack_response({{:ok, _}, conn}, state) do
+    schedule_fetch(%{state | conn: conn, worker_state: :ok, job_ref: nil, job_id: nil})
+  end
+
+  defp handle_ack_response({{:error, _}, conn}, state) do
+    # todo: need to work out what we should do here
```

Comment:
Is this something that will be logged? or something else needs to happen?

Reply:
If the job has successfully completed and this happens it currently will never send an `ACK` meaning faktory will retry the job after the `reserve_for` has expired. We definitely need to try and handle this but wasn't sure on the best approach.